### PR TITLE
apt: include arch in check for installed packages on multi-arch systems

### LINKED ISSUE #24673: apt deb: with url always reports changed
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

apt deb (with url)

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### CONFIGURATION

unchanged

##### OS / ENVIRONMENT

Distributor ID: neon
Description:    KDE neon User Edition 5.9
Release:        16.04
Codename:       xenial

##### SUMMARY

I have an apt-deb task which always reports "changed" although the deb is already installed.

##### STEPS TO REPRODUCE

```yaml
# main.yml: Install Drivers for Brother DCP-195C
# See:
# http://support.brother.com/g/b/downloadlist.aspx?c=de&lang=de&prod=dcp195c_all&os=128&flang=English

- hosts: localhost
  tasks:
    - name: "brother | install printer driver"
      become: yes
      ignore_errors: yes
      apt:
        deb: "{{ item }}"
      with_items:
        - "http://www.brother.com/pub/bsc/linux/packages/dcp195clpr-1.1.3-1.i386.deb" # lpr before cupswrapper
        - "http://www.brother.com/pub/bsc/linux/packages/dcp195ccupswrapper-1.1.3-1.i386.deb"
```
Then run `ansible-playbook -i localhost, -c local -K -vvvv main.yml`

##### EXPECTED RESULTS

Ansible should report no changes since I have already run this task multiple times. The packages are already installed.

##### ACTUAL RESULTS

Task always reports "changed".

```
$ ansible-playbook -i localhost, -c local -K -vvvv main.yml
Using /etc/ansible/ansible.cfg as config file
SUDO password: 
Set default localhost to localhost
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.7/dist-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: main.yml ********************************************************************************************************************************************************************
1 plays in main.yml

PLAY [localhost] **********************************************************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************************************************
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/system/setup.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: stefan
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/stefan/.ansible/tmp/ansible-tmp-1494945861.24-188838315268741 `" && echo ansible-tmp-1494945861.24-188838315268741="` echo /home/stefan/.ansible/tmp/ansible-tmp-1494945861.24-188838315268741 `" ) && sleep 0'
<localhost> PUT /tmp/tmpsz1R_v TO /home/stefan/.ansible/tmp/ansible-tmp-1494945861.24-188838315268741/setup.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/stefan/.ansible/tmp/ansible-tmp-1494945861.24-188838315268741/ /home/stefan/.ansible/tmp/ansible-tmp-1494945861.24-188838315268741/setup.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/stefan/.ansible/tmp/ansible-tmp-1494945861.24-188838315268741/setup.py; rm -rf "/home/stefan/.ansible/tmp/ansible-tmp-1494945861.24-188838315268741/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [brother | install printer driver] ***********************************************************************************************************************************************
task path: /home/stefan/ansible-test/main.yml:7
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/packaging/os/apt.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: stefan
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/stefan/.ansible/tmp/ansible-tmp-1494945861.7-210324426232444 `" && echo ansible-tmp-1494945861.7-210324426232444="` echo /home/stefan/.ansible/tmp/ansible-tmp-1494945861.7-210324426232444 `" ) && sleep 0'
<localhost> PUT /tmp/tmppjkF7q TO /home/stefan/.ansible/tmp/ansible-tmp-1494945861.7-210324426232444/apt.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/stefan/.ansible/tmp/ansible-tmp-1494945861.7-210324426232444/ /home/stefan/.ansible/tmp/ansible-tmp-1494945861.7-210324426232444/apt.py && sleep 0'
<localhost> EXEC /bin/sh -c 'sudo -H -S  -p "[sudo via ansible, key=vcnveohugtpycwxqhljqvbesjsyjnwlj] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-vcnveohugtpycwxqhljqvbesjsyjnwlj; /usr/bin/python /home/stefan/.ansible/tmp/ansible-tmp-1494945861.7-210324426232444/apt.py; rm -rf "/home/stefan/.ansible/tmp/ansible-tmp-1494945861.7-210324426232444/" > /dev/null 2>&1'"'"' && sleep 0'                                                                                                                                                            
changed: [localhost] => (item=http://www.brother.com/pub/bsc/linux/packages/dcp195clpr-1.1.3-1.i386.deb) => {
    "changed": true, 
    "diff": {
        "prepared": "(Reading database ... 191206 files and directories currently installed.)\nPreparing to unpack .../dcp195clpr-1.1.3-1.i386.deb ...\nUnpacking dcp195clpr:i386 (1.1.3-1) over (1.1.3-1) ...\nSetting up dcp195clpr:i386 (1.1.3-1) ..."                                                                                                                     
    }, 
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false, 
            "autoremove": null, 
            "cache_valid_time": 0, 
            "deb": "/tmp/ansible_UkG3fk/dcp195clpr-1.1.3-1.i386.deb", 
            "default_release": null, 
            "dpkg_options": "force-confdef,force-confold", 
            "force": false, 
            "install_recommends": null, 
            "only_upgrade": false, 
            "package": null, 
            "purge": false, 
            "state": "present", 
            "update_cache": null, 
            "upgrade": null
        }
    }, 
    "item": "http://www.brother.com/pub/bsc/linux/packages/dcp195clpr-1.1.3-1.i386.deb", 
    "stderr": "mkdir: cannot create directory '/var/spool/lpd/dcp195c': No such file or directory\nchown: cannot access '/var/spool/lpd/dcp195c': No such file or directory\nchgrp: cannot access '/var/spool/lpd/dcp195c': No such file or directory\nchmod: cannot access '/var/spool/lpd/dcp195c': No such file or directory\n",                                           
    "stderr_lines": [
        "mkdir: cannot create directory '/var/spool/lpd/dcp195c': No such file or directory", 
        "chown: cannot access '/var/spool/lpd/dcp195c': No such file or directory", 
        "chgrp: cannot access '/var/spool/lpd/dcp195c': No such file or directory", 
        "chmod: cannot access '/var/spool/lpd/dcp195c': No such file or directory"
    ], 
    "stdout": "(Reading database ... 191206 files and directories currently installed.)\nPreparing to unpack .../dcp195clpr-1.1.3-1.i386.deb ...\nUnpacking dcp195clpr:i386 (1.1.3-1) over (1.1.3-1) ...\nSetting up dcp195clpr:i386 (1.1.3-1) ...\n",                                                                                                                        
    "stdout_lines": [
        "(Reading database ... 191206 files and directories currently installed.)", 
        "Preparing to unpack .../dcp195clpr-1.1.3-1.i386.deb ...", 
        "Unpacking dcp195clpr:i386 (1.1.3-1) over (1.1.3-1) ...", 
        "Setting up dcp195clpr:i386 (1.1.3-1) ..."
    ]
}
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/packaging/os/apt.py
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/stefan/.ansible/tmp/ansible-tmp-1494945866.04-121364415219253 `" && echo ansible-tmp-1494945866.04-121364415219253="` echo /home/stefan/.ansible/tmp/ansible-tmp-1494945866.04-121364415219253 `" ) && sleep 0'
<localhost> PUT /tmp/tmp6N1eXt TO /home/stefan/.ansible/tmp/ansible-tmp-1494945866.04-121364415219253/apt.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/stefan/.ansible/tmp/ansible-tmp-1494945866.04-121364415219253/ /home/stefan/.ansible/tmp/ansible-tmp-1494945866.04-121364415219253/apt.py && sleep 0'
<localhost> EXEC /bin/sh -c 'sudo -H -S  -p "[sudo via ansible, key=ruiizwouxuptmgkbuidmglnpjicccykm] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-ruiizwouxuptmgkbuidmglnpjicccykm; /usr/bin/python /home/stefan/.ansible/tmp/ansible-tmp-1494945866.04-121364415219253/apt.py; rm -rf "/home/stefan/.ansible/tmp/ansible-tmp-1494945866.04-121364415219253/" > /dev/null 2>&1'"'"' && sleep 0'
changed: [localhost] => (item=http://www.brother.com/pub/bsc/linux/packages/dcp195ccupswrapper-1.1.3-1.i386.deb) => {
    "changed": true, 
    "diff": {
        "prepared": "(Reading database ... 191206 files and directories currently installed.)\nPreparing to unpack .../dcp195ccupswrapper-1.1.3-1.i386.deb ...\nRestarting cups (via systemctl): cups.service.\nUnpacking dcp195ccupswrapper:i386 (1.1.3-1) over (1.1.3-1) ...\nSetting up dcp195ccupswrapper:i386 (1.1.3-1) ...\nRestarting cups (via systemctl): cups.service."
    }, 
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false, 
            "autoremove": null, 
            "cache_valid_time": 0, 
            "deb": "/tmp/ansible_Jj_E9s/dcp195ccupswrapper-1.1.3-1.i386.deb", 
            "default_release": null, 
            "dpkg_options": "force-confdef,force-confold", 
            "force": false, 
            "install_recommends": null, 
            "only_upgrade": false, 
            "package": null, 
            "purge": false, 
            "state": "present", 
            "update_cache": null, 
            "upgrade": null
        }
    }, 
    "item": "http://www.brother.com/pub/bsc/linux/packages/dcp195ccupswrapper-1.1.3-1.i386.deb", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "(Reading database ... 191206 files and directories currently installed.)\nPreparing to unpack .../dcp195ccupswrapper-1.1.3-1.i386.deb ...\nRestarting cups (via systemctl): cups.service.\nUnpacking dcp195ccupswrapper:i386 (1.1.3-1) over (1.1.3-1) ...\nSetting up dcp195ccupswrapper:i386 (1.1.3-1) ...\nRestarting cups (via systemctl): cups.service.\n", 
    "stdout_lines": [
        "(Reading database ... 191206 files and directories currently installed.)", 
        "Preparing to unpack .../dcp195ccupswrapper-1.1.3-1.i386.deb ...", 
        "Restarting cups (via systemctl): cups.service.", 
        "Unpacking dcp195ccupswrapper:i386 (1.1.3-1) over (1.1.3-1) ...", 
        "Setting up dcp195ccupswrapper:i386 (1.1.3-1) ...", 
        "Restarting cups (via systemctl): cups.service."
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP ****************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
```

tag: @evgeni

### PR DESCRIPTION
**Pull request by evgeni**
##### SUMMARY
See #24673 and #24703.

On systems that have multi-arch enabled, you need to lookup packages via `package:arch` from `apt.Cache()`. Newer versions of `python-apt` also accept this notation when multi-arch is not enabled, but the version shipped in Debian Jessie and Ubuntu 14.04 does not.

Thanks: Stefan Löwen <stefan.loewen@gmail.com>

Fixes: #24673 
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
apt

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 7563d93901) last updated 2017/05/19 23:17:35 (GMT +200)
```

### CODE REVIEW

File: test/integration/targets/apt/tasks/apt.yml
Comment (pilou-) [inline test/integration/targets/apt/tasks/apt.yml]: You need to check that `{{ apt_foreign_arch }}` isn't the default architecture.

File: test/integration/targets/apt/tasks/apt.yml
Comment (pilou-) [inline test/integration/targets/apt/tasks/apt.yml]: You could check results of previous tasks  `install hello` and `uninstall hello` too.

File: test/integration/targets/apt/tasks/apt.yml
Comment (pilou-) [inline test/integration/targets/apt/tasks/apt.yml]: The cleaning tasks need to be executed on an [`always`](http://docs.ansible.com/ansible/playbooks_blocks.html#error-handling) block.

File: test/integration/targets/apt/tasks/apt.yml
Comment (pilou-) [inline test/integration/targets/apt/tasks/apt.yml]: In order to improve readability, test tasks related to multi-arch could be moved in a dedicated YAML file and used here using an `include` task.

**Review by pilou- — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
Number of failures in the integration tests

File: test/integration/targets/apt/tasks/apt.yml
Comment (pilou-) [inline test/integration/targets/apt/tasks/apt.yml]: This task fails due to #23601 when the control machine isn't using Debian (or a derivatives).

Besides, when Debian is used on the control machine, following error appears:

```
TASK [apt : remove second architecture] *********************************************************************************************************************
fatal: [testhost]: FAILED! => {"changed": true, "cmd": ["dpkg", "--remove-architecture", "i386"], "delta": "0:00:00.015897", "end": "2017-05-21 21:55:06.487061", "failed": true, "rc": 2, "start": "2017-05-21 21:55:06.471164", "stderr": "dpkg: error: cannot remove architecture 'i386' currently in use by the database", "stderr_lines": ["dpkg: error: cannot remove architecture 'i386' currently in use by the database"], "stdout": "", "stdout_lines": []}
```

You could use instead something like:
```
 - name: Remove packages with foreign architecture
   apt:
     name: "*:{{ apt_foreign_arch }}"
     state: absent
     purge: yes
```

File: lib/ansible/modules/packaging/os/apt.py
Comment (mikedlr) [inline lib/ansible/modules/packaging/os/apt.py:560]: Elsewhere in this module the cache which is returned from `get_cache(module)` is always carefully used.  In this case there is a direct call to apt.Cache().  Do you know why there is this discrepancy and is it safe?  If possible it would be preferable to use the cache object which is already passed into this function as an argument.

File: lib/ansible/modules/packaging/os/apt.py
Comment (mikedlr) [inline lib/ansible/modules/packaging/os/apt.py:551]: This is a huge try/except block which makes debugging what's going on quite difficult and can mask nasty errors.  This and other prolbems with this code are from before the current PR, however it would be really good if this situation could be improved.

File: test/integration/targets/apt/tasks/apt.yml
Comment (evgeni) [reply test/integration/targets/apt/tasks/apt.yml]: ack, did that (not pushed yet), works fine, thanks

File: test/integration/targets/apt/tasks/apt.yml
Comment (evgeni) [reply test/integration/targets/apt/tasks/apt.yml]: why? the tests are marked as destructive, and having the second arch enabled does not really hurt. on the other side blocks make playbooks less readable.

**Comment by evgeni** [comment]
ready_for_review

**Review by mikedlr — COMMENTED** [review]
Hi @evgeni ;  I think this is fine as a small fix and reasonable to go forward.  You could close this as is.

Talking generally (not just this specific patch) the module should have a bit of work done on it;

- the module currently calls dpkg three times for every single .deb it queries - this could be replaced with one call querying 
- the dpg call should return a simple data structure which can be queried (like the cache??) which could then be mocked and tested easily.
- it would be good to have more coverage in the test cases for package upgrades [which are currently missing from the tests (https://codecov.io/gh/ansible/ansible/src/e38c29e71907e273b14b55c60e2899254949a2b2/lib/ansible/modules/packaging/os/apt.py#L660) 

If I was refactoring, I think I would pull out the following functions which would be easy to unit test:

```
def read_debs(list_of_files):
    # blah blah
   available_packages = (# equivalent of 
   for i in *; 
       do dpkg-deb --showformat '${PACKAGE} ${ARCHITECTURE} ${VERSION}\n' --show $i ; 
    done ) 
    return 
```

```
def identify_changes(installed_packages, available_packages):
    # blah blah
    return [list of packages where available is better]
```

And I'd probably pull out a Package object for both installed and available packages which would have

- package.version()
- package.architecture()
- package.name()

which you can do a comparison on but would could run `package.path()` for the installable debs so you could easily build the list of packages to install  

I met with @gundalow yesterday and we were discussing this.  If you want to take it these you could do it together with the rest of the patch.  If not then one of us will bring this up at the core team meeting and hopefully someone else would look into it.

**Comment by evgeni** [comment]
HI @mikedlr, I totally agree that the module could need some good love, refactoring and tests. I tried to be as non-invasive as possible in my patch, as I did not want to take the refactoring route just now and I have no clue why some parts (like the Cache you mention) are written like they are today.

I would favor we would fix that bug (which it is) now, and then decide on a separate plan how to get the module in a nicer shape. Happy to help with the further tasks, but probably not as the main driver (time constraints etc).

**Comment by mikedlr** [comment]
Okay, then please don't merge the [unit test I made](https://github.com/mikedlr/ansible/commit/25dbdacbe8ad83b62d370c321d7d6afdc22cab7c) right now.  It will be brittle under the refactoring.  Whoever does the refactoring should either merge it or add new separate unit tests.

**Comment by mikedlr** [comment]
+1

**Comment by evgeni** [comment]
OK, backed out your commit.

**Pull request merged**
